File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/IngestionMapping.java
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package com.microsoft.azure.kusto.ingest;

import java.util.Arrays;

/**
 * This class describes the ingestion mapping to use for an ingestion request.
 * When a CSV data source schema and the target schema doesn't match or when using JSON, AVRO formats,
 * there is a need to define an ingestion mapping to map the source schema to the table schema.
 * This class describes a pre-defined ingestion mapping by its name- mapping reference and its kind.
 */
public class IngestionMapping {
    private ColumnMapping[] columnMappings;
    private IngestionMappingKind ingestionMappingKind;
    private String ingestionMappingReference;

    /**
     * Creates a default ingestion mapping with null kind and empty mapping reference.
     */
    public IngestionMapping() {
    }

    /**
     * Creates an ingestion mapping with the given parameters.
     *
     * @param ingestionMappingReference String: the name of the pre-defined ingestion mapping.
     * @param ingestionMappingKind      IngestionMappingKind: the format of the source data to map from.
     */
    public IngestionMapping(String ingestionMappingReference, IngestionMappingKind ingestionMappingKind) {
        this.ingestionMappingReference = ingestionMappingReference;
        this.ingestionMappingKind = ingestionMappingKind;
    }

    /**
     * Please use setIngestionMappingReference for production as passing the mapping every time is wasteful
     *
     * @param columnMappings       Array of columnMappings of the same kind.
     * @param ingestionMappingKind IngestionMappingKind: the format of the source data to map from.
     */
    public IngestionMapping(ColumnMapping[] columnMappings, IngestionMappingKind ingestionMappingKind) {
        this.columnMappings = columnMappings;
        this.ingestionMappingKind = ingestionMappingKind;
    }

    /**
     * Copy constructor for IngestionMapping.
     *
     * @param other the instance to copy from
     */
    public IngestionMapping(IngestionMapping other) {
        this.ingestionMappingKind = other.ingestionMappingKind;
        this.ingestionMappingReference = other.ingestionMappingReference;
        if (other.columnMappings != null) {
            this.columnMappings = Arrays.stream(other.columnMappings).map(ColumnMapping::new).toArray(ColumnMapping[]::new);
        }
    }

    /**
     * Sets the ingestion mapping reference parameters
     *
     * @param ingestionMappingReference String: the name of the pre-defined ingestion mapping.
     * @param ingestionMappingKind      IngestionMappingKind: the format of the source data to map from.
     */
    public void setIngestionMappingReference(String ingestionMappingReference, IngestionMappingKind ingestionMappingKind) {
        this.ingestionMappingReference = ingestionMappingReference;
        this.ingestionMappingKind = ingestionMappingKind;
    }

    /**
     * Please use setIngestionMappingReference for production as passing the mapping every time is wasteful
     * Sets the ingestion mapping parameters
     *
     * @param columnMappings       Array of columnMappings of the same kind.
     * @param ingestionMappingKind IngestionMappingKind: the format of the source data to map from.
     */
    public void setIngestionMapping(ColumnMapping[] columnMappings, IngestionMappingKind ingestionMappingKind) {
        this.columnMappings = columnMappings;
        this.ingestionMappingKind = ingestionMappingKind;
    }

    public IngestionMappingKind getIngestionMappingKind() {
        return ingestionMappingKind;
    }

    public String getIngestionMappingReference() {
        return ingestionMappingReference;
    }

    public ColumnMapping[] getColumnMappings() {
        return columnMappings;
    }

    /*
     * Represents an ingestion mapping kind - the format of the source data to map from.
     */
    public enum IngestionMappingKind {
        CSV("Csv"),
        JSON("Json"),
        AVRO("Avro"),
        PARQUET("Parquet"),
        SSTREAM("SStream"),
        ORC("Orc"),
        APACHEAVRO("ApacheAvro"),
        W3CLOGFILE("W3CLogFile");

        private final String kustoValue;

        IngestionMappingKind(String kustoValue) {
            this.kustoValue = kustoValue;
        }

        public String getKustoValue() {
            return kustoValue;
        }
    }
}

File: ingest-v2/src/main/kotlin/com/microsoft/azure/kusto/ingest/v2/common/models/mapping/IngestionMapping.kt
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
package com.microsoft.azure.kusto.ingest.v2.common.models.mapping

import com.microsoft.azure.kusto.ingest.v2.models.Format
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlin.collections.emptyList
import kotlinx.serialization.Serializable as KSerializable

@KSerializable
class IngestionMapping
private constructor(
    val ingestionMappingReference: String,
    val columnMappings: List<ColumnMapping>,
    val ingestionMappingType: IngestionMappingType,
) {
    constructor(
        other: IngestionMapping,
    ) : this(
        other.ingestionMappingReference,
        other.columnMappings.map {
            ColumnMapping(
                it.columnName,
                columnType = it.columnType,
                properties = it.properties,
            )
        },
        other.ingestionMappingType,
    )

    constructor(
        ingestionMappingReference: String,
        ingestionMappingType: IngestionMappingType,
    ) : this(
        columnMappings = emptyList(),
        ingestionMappingReference = ingestionMappingReference,
        ingestionMappingType = ingestionMappingType,
    )

    constructor(
        columnMappings: List<ColumnMapping>,
        ingestionMappingType: IngestionMappingType,
    ) : this(
        columnMappings = columnMappings,
        ingestionMappingReference = "",
        ingestionMappingType = ingestionMappingType,
    )

    /**
     * Serializes the column mappings to a JSON string representation.
     *
     * @return JSON string representation of the column mappings
     */
    fun serializeColumnMappingsToJson(): String {
        return Json.encodeToString(columnMappings)
    }

    enum class IngestionMappingType(
        val kustoValue: String,
        val format: Format,
    ) {
        CSV("Csv", Format.csv),
        JSON("Json", Format.json),
        AVRO("Avro", Format.avro),
        PARQUET("Parquet", Format.parquet),
        SSTREAM("SStream", Format.sstream),
        ORC("Orc", Format.orc),
        APACHEAVRO("ApacheAvro", Format.apacheavro),
        W3CLOGFILE("W3CLogFile", Format.w3clogfile),
    }
}
